Firestore/Swift/Source/Codable/TimestampDecodingStrategy.swift

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFirestore
+import FirebaseSharedSwift
+
+@available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *)
+private var _iso8601Formatter: ISO8601DateFormatter = {
+  let formatter = ISO8601DateFormatter()
+  formatter.formatOptions = .withInternetDateTime
+  return formatter
+}()
+
+extension StructureDecoder.DateDecodingStrategy {
+  public static func timestamp(fallback: StructureDecoder
+    .DateDecodingStrategy = .deferredToDate) -> StructureDecoder.DateDecodingStrategy {
```

Comment:
Can you quickly explain when this would be used in Firestore since we do not use JSON-formatted dates?

Reply:
Sure - no other reason than: someone might have a different preference then encoding to / decoding from the `Timestamp` type.

If I wish to represent my date as the number of seconds since epoch or as a string specifying week-year and week of the date: "2021-44" or as ISO 8601 or anything else, then I don't think that the API should force me to use `Timestamp`.

If you disagree then this can of course just be removed.

Reply:
If I understand this correctly, this allows user to use String-based dates in their Codable objects, but we will store them as Timestamps? If so, then this makes sense.

Reply:
As this is in the direction of decoding it's more like:
If you have a `Date` in you encodable model and a `Timestamp` in you Firestore document, then the decoding will map the `Timestamp` to the `Date`. This step is to be API compatible with functionality that is already a part of the existing `Firestore.Decoder`. But it also lets you express a date decoding strategy to use when the Firestore document is _not_ a `Timestamp`.
Perhaps you might always store iso8601 encoded date strings in your Firestore document - or you might be storing seconds or miliseconds since epoch. For whatever reasons - historical, compatibility with another framework, etc., etc.
So this just allows the exact same configurability that users may be used to from `JSONDecoder` while still being compatible with the existing internals of `Firestore.Decoder` that would always automatically decode a `Timestamp` to a `Date` if that was requested by the user.

This strategy should however be internal instead of public since it's used internally and is just there to ensure backwards compatibility with the existing `Firestore.Decoder`.